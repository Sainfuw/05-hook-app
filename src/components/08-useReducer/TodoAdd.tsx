import React from "react";
import { useForm } from "../../hooks/useForm";
import { Todo } from "../../types/appTypes";

interface Props {
  handleAddTodo: (newTodo: Todo) => void;
}

export const TodoAdd = ({ handleAddTodo }: Props) => {
  const { form, handleInputChange, reset } = useForm({ description: "" });
  const { description } = form;

  const handleSubmit = (e: any) => {
    e.preventDefault();
    if (description.trim().length < 2) return;
    const newTodo: Todo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="Aprender ..."
          autoComplete="off"
          className="form-control"
          value={description}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-outline-primary mt-2 w-100">
          Agregar
        </button>
      </form>
    </>
  );
};
