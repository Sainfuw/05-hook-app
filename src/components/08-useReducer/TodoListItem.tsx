import React from "react";
import { Todo } from "../../types/appTypes";

interface Props {
  todo: Todo;
  index: number;
  handleDelete: (id: number) => void;
  toggleDone: (id: number) => void;
}

export const TodoListItem = ({
  todo,
  index,
  handleDelete,
  toggleDone,
}: Props) => {
  return (
    <li className="list-group-item">
      <p
        className={`${todo.done && " complete"}`}
        onClick={() => toggleDone(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-outline-danger"
        onClick={() => handleDelete(todo.id)}
      >
        Borrar
      </button>
    </li>
  );
};
