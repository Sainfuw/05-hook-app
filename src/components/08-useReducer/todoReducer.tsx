import { Todo } from "../../types/appTypes";

export type TodoAction =
  | { type: "add"; payload: Todo }
  | { type: "toggleDone"; payload: number }
  | { type: "remove"; payload: number };

export const todoReducer = (state: Todo[], action: TodoAction) => {
  switch (action.type) {
    case "add":
      return [...state, action.payload];

    case "remove":
      return state.filter((todo) => todo.id !== action.payload);

    case "toggleDone":
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      );

    default:
      return state;
  }
};
