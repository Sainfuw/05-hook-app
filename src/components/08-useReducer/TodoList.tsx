import React from "react";
import { Todo } from "../../types/appTypes";
import { TodoListItem } from "./TodoListItem";

interface Props {
  todos: Todo[];
  handleDelete: (id: number) => void;
  toggleDone: (id: number) => void;
}

export const TodoList = ({ todos, handleDelete, toggleDone }: Props) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((todo, index) => (
        <TodoListItem
          key={todo.id}
          todo={todo}
          index={index}
          handleDelete={handleDelete}
          toggleDone={toggleDone}
        />
      ))}
    </ul>
  );
};
