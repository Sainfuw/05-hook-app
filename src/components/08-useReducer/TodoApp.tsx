import React, { useReducer, useEffect } from "react";
import "./styles.css";
import { todoReducer } from "./todoReducer";
import { Todo } from "../../types/appTypes";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

const init = () => {
  const storedTodos = localStorage.getItem("todos");
  return typeof storedTodos === "string" ? JSON.parse(storedTodos) : [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const handleAddTodo = (newTodo: Todo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  const handleDelete = (id: number) => {
    dispatch({
      type: "remove",
      payload: id,
    });
  };

  const toggleDone = (id: number) => {
    dispatch({
      type: "toggleDone",
      payload: id,
    });
  };

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  return (
    <>
      <h4 id="TodoAppTitle">TodoApp ({todos.length})</h4>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            toggleDone={toggleDone}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};
