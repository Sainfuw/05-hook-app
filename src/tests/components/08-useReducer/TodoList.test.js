import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { demoTodos } from "../../fixtures/demoTodos";
import { TodoList } from "../../../components/08-useReducer/TodoList";

describe("Pruebas en TodoList.test", () => {
  const handleDelete = jest.fn();
  const toggleDone = jest.fn();
  const wrapper = shallow(
    <TodoList
      todos={demoTodos}
      handleDelete={handleDelete}
      toggleDone={toggleDone}
    />
  );

  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe tener dos TodoListItem", () => {
    expect(wrapper.find("TodoListItem").length).toBe(demoTodos.length);
    expect(wrapper.find("TodoListItem").at(0).prop("handleDelete")).toEqual(
      expect.any(Function)
    );
  });
});
