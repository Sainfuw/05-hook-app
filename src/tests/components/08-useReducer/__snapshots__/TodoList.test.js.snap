// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoList.test debe mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
    toggleDone={[MockFunction]}
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
    toggleDone={[MockFunction]}
  />
</ul>
`;
